lab_02_review/inc/matrix.h

```diff
@@ -0,0 +1,125 @@
+#ifndef MATRIX_H
+#define MATRIX_H
+#include "base_matrix.h"
+#include "concept.h"
+#include "const_iterator.h"
+#include "iterator.h"
+#include <initializer_list>
+#include <memory>
+
+template <NumType T> class Matrix : public BaseMatrix
+{
+public:
+    friend Iterator<T>;
```

Comment:
не нужно

Reply:
Ответьте, пожалуйста, если будет время, есть проблемы с пониманием идеи о ненужности дружбы.

Согласно [документации](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this) `enable_shared_from_this` позволяет объекту t,  _**который в настоящее время управляется `std::shared_ptr`**_ с именем pt, безопасно генерировать дополнительные `std::shared_ptr` экземпляры pt1, pt2, ... которые совместно владеют t с pt. 

То есть для того, чтобы избавиться от дружбы вашим способом, если я правильно вас понимаю, мне надо  запретить пользователю библиотеки создавать сами объекты и заставить его создавать только `shared_ptr` на объекты явным вызовом обертки над конструктором, как ниже. Потому что сделать из указателя `this` `shared_ptr` без уже управляемого  `std::shared_ptr`ом объекта нельзя.
```
#include <iostream>
#include <memory>

class Best : public std::enable_shared_from_this<Best>
{
    struct Private{};
 
public:
    // Constructor is only usable by this class
    Best(Private) {}
 
    // Everyone else has to use this factory function
    // Hence all Best objects will be contained in shared_ptr
    static std::shared_ptr<Best> create()
    {
        return std::make_shared<Best>(Private());
    }
 
    std::shared_ptr<Best> getptr()
    {
        return shared_from_this();
    }
};

void testBest()
{
    std::shared_ptr<Best> best0 = Best::create();
    std::shared_ptr<Best> best1 = best0->getptr();
    std::cout << "best1.use_count() = " << best1.use_count() << '\n';
 
    // Best stackBest; // <- Will not compile because Best::Best() is private.
}
 
int main()
{
    testBest();
}
```
Правильно ли я понимаю вашу идею? Если да, то почему такое ограничение пользователя лучше, чем дружба? 

Можно ли оставить дружбу и не переписывать код подобным образом?

Reply:
да, действительно. пока предлагаю следующее: сделайте у итераторы два конструкртора:
1. принимает weak_ptr на память под матрицу и используется для создания итератора из матрицы
2. принимает weak_ptr на матрицу и используется внешними пользователями

Reply:
я там в вашей группе еще паре человек сказал сделать enable_shared_from_this можете рассказать им об этой проблеме

Reply:
Спасибо, посмотрю, и сообщу ребятам.

Reply:
К сожалению, у нас всё ещё не вышло понять вашу новую идею. Можете, пожалуйста, объяснить подробнее, что вы имеете в виду.